lib/opal/rewriters/logical_operator_assignment.rb

```diff
@@ -0,0 +1,199 @@
+require 'opal/rewriters/base'
+
+module Opal
+  module Rewriters
+    class LogicalOperatorAssignment < Base
+      class AbstractHandler < self
+        def self.get(lhs, rhs)
+          raise NotImplementedError
+        end
+
+        def self.set(lhs, rhs)
+          raise NotImplementedError
+        end
+
+        def self.process(lhs, rhs, root_type)
+          s(root_type,
+            get(lhs, rhs),
+            set(lhs, rhs))
+        end
+
+        def self.new_temp
+          @@counter ||= 0
+          @@counter += 1
+          :"$or_equals_lhs_recvr_tmp_#{@@counter}"
+        end
+
+        def self.reset_tmp_counter!
+          @@counter = 0
+        end
+      end
+
+      # Takes    `lhs ||= rhs`
+      # Produces `lhs || lhs = rhs`
+      class LocalVariableHandler < AbstractHandler
+        def self.get(lhs, rhs)
+          lhs.updated(:lvar) # lhs
+        end
+
+        def self.set(lhs, rhs)
+          lhs.updated(:lvasgn, [*lhs, rhs]) # lhs = rhs
+        end
+      end
```

Comment:
@iliabylich maybe this series of classes can be simplified/shortened with this kind of code:

```ruby
class GetSetHandler < AbstractHandler
  def self.get(lhs, rhs)
    lhs.updated(@get_type) # lhs
  end
  def self.set(lhs, rhs)
    lhs.updated(@set_type, [*lhs, rhs]) # lhs = rhs
  end
end

# Takes    `lhs ||= rhs`
# Produces `lhs || lhs = rhs`
class LocalVariableHandler < AbstractHandler
  @get_type = :lvar
  @set_type = :lvasgn
end
```

or even (maybe too far?)

```ruby
class GetSetHandler < AbstractHandler
  def self.get(lhs, rhs)
    lhs.updated(@get_type) # lhs
  end
  def self.set(lhs, rhs)
    lhs.updated(@set_type, [*lhs, rhs]) # lhs = rhs
  end
  
  def self.handle(set_type, get_type)
    @set_type = set_type
    @get_type = get_type
    HANDLERS[set_type] = self
  end
end

# Takes    `lhs ||= rhs`
# Produces `lhs || lhs = rhs`
class LocalVariableHandler < AbstractHandler
  handle :lvasgn, :lvar
end
```

Anyway this should be both simpler and faster (more optimizable) and take less memory (I'm thinking about the JS version mostly)